cpp/src/io/csv/reader_impl.cu

```diff
@@ -346,194 +351,215 @@ table_with_metadata reader::impl::read(size_t range_offset,
-      header.assign(h_data + *(it_begin), h_data + *(it_begin + 1));
+      if (num_rows > 0 && num_row_offsets > static_cast<size_t>(num_rows)) {
+        // Got the desired number of rows
+        break;
+      }
     } else {
-      header.assign(h_data + *(it_begin + args_.header), h_data + *(it_begin + args_.header + 1));
-      it_begin += args_.header + 1;
+      // Discard data (all rows below skip_rows), keeping one character for history
+      size_t discard_bytes = std::max(data_.size(), sizeof(char)) - sizeof(char);
+      if (discard_bytes != 0) {
+        data_.erase(data_.begin(), data_.begin() + discard_bytes);
+        buffer_pos += discard_bytes;
+      }
     }
-  }
-
-  // Exclude the rows that exceed past the requested number
-  if (num_rows >= 0 && num_rows < std::distance(it_begin, it_end)) {
-    it_end = it_begin + num_rows + 1;
-  }
-
-  // Exclude the rows that are to be skipped from the end
-  if (skip_end_rows > 0 && skip_end_rows < std::distance(it_begin, it_end)) {
-    it_end -= skip_end_rows;
-  }
-
-  const uint64_t offset_start = *it_begin;
-  const uint64_t offset_end   = *(it_end - 1);
+    pos = target_pos;
+  } while (pos < h_size);
+}
 
-  // Copy out the row starts to use for row-column data parsing
-  if (offset_start != offset_end) {
-    if (offset_start != 0) {
-      for (auto it = it_begin; it != it_end; ++it) { *it -= offset_start; }
+size_t reader::impl::parse_csv_header(const char *h_data, size_t h_size)
+{
+  using namespace cudf::io::csv::gpu;
+  size_t header_start = 0;
+  size_t header_end   = 0;
+  size_t data_start   = 0;
+  cudf::size_type row = 0;
+  int ctx             = ROW_CTX_NONE;
+  int commentchar     = (opts.comment != '\0') ? opts.comment : 0x100;
+  int quotechar       = (opts.quotechar != '\0') ? opts.quotechar : 0x100;
+  int escapechar      = 0x100;  // FIXME: Add escapechar to API
+  /// Escapable quote-terminating delimiter (must have escapechar support)
+  int esc_delimiter = 0x100;
+  int prev_ch       = opts.terminator;
+
+  for (size_t pos = 0; pos < h_size; pos++) {
+    int ch = h_data[pos];
+    if ((ch == opts.terminator && ctx != ROW_CTX_QUOTE) || pos + 1 == h_size) {
+      if (ctx == ROW_CTX_COMMENT) {
+        ctx = ROW_CTX_NONE;
+      } else if (opts.skipblanklines &&
+                 (pos == header_start ||
+                  (pos == header_start + 1 && opts.terminator == '\n' && prev_ch == '\r'))) {
+      } else if (row >= args_.header || pos + 1 == h_size) {
+        header_end = pos + 1;
+        if (row == args_.header) { data_start = header_end; }
+        break;
+      } else {
+        row++;
+      }
+      header_start = header_end = pos + 1;
+    } else if (ctx == ROW_CTX_NONE) {
+      if (prev_ch == opts.terminator && ch == commentchar) {
+        ctx = ROW_CTX_COMMENT;
+      } else if (ch == quotechar && (prev_ch == opts.terminator || prev_ch == opts.delimiter ||
+                                     prev_ch == quotechar)) {
```

Comment:
Yeah. That's one illegal case where we would likely fail differently than pandas (doublequotes are only allowed in quoted strings or if quoting is disabled). Not planning to handle this particular case in this PR (the list of corner-cases is almost never-ending).
At some point soon, we'll need to deal with pandas compatibility for both the row-parsing stage and the data conversion stage (we should be able to match pandas in most cases ~ for the row parser, this potentially means looking at a larger window besides just ch_prev).

Reply:
Fair enough; I was just looking for ways to break the parser, but this would be illegal input as you say. Anyway, I suppose it would be easier to throw an exception when a quote is encountered and the previous char is _not_ a terminator, delimiter, or quote (which would fail at the first quote in `foo"",bar`). Assuming you want to fail when invalid input is encountered, rather than "garbage in garbage out", of course.

Reply:
The correct behavior would be to match pandas (which might be hard in the case of column names). You often can't rule out invalid until you parse all the way to the end of the file (or give up after seeing a column name with a length greater than one million characters)

Reply:
[For that particular case, I think the simplest solution would be to add a lookahead of the next character, so we don't change the quoting state when encountering double-quotes]

Reply:
Will mark this as resolved for now: if we hit issues with ignoring quotes that do not start immediately after the delimiter, we could easily revert to the simple 2-state quoting scheme in the previous implementation by simply commenting out the delimiter-before-opening-quote check (the good thing about this new implementation is that it makes it possible to add more features to the parser while remaining fully data-parallel). I think it should be possible to have a data-parallel parser that fully matches pandas behavior (also hoping this is doable without adding more parser states ~ maybe a couple of flags/sub-states that would only impact the first state transition, or just brute-force look-ahead/look-behind for rare corner cases).